flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/deserializer/PulsarDeserializationSchema.java

```diff
@@ -67,7 +67,13 @@ default void open(InitializationContext context) throws Exception {
      * @param message The message decoded by pulsar.
      * @param out The collector to put the resulting messages.
      */
-    void deserialize(Message<byte[]> message, Collector<T> out) throws Exception;
+    void deserialize(Message<?> message, Collector<T> out) throws Exception;
+
+    /** @return The related Pulsar Schema for this serializer. */
+    default Schema<T> schema() {
```

Comment:
I have a question of the control flow here. We have two default options, return null or throw Exception. Using exception and try catch to  distinguish flink schema and pulsar schema is kind like "Exception as control flow". 

Returning null is not good as well.

Reply:
This is quite easy to answer. We would use this exception as the beginning of the workflow. This is a fail-fast strategy for validating if this schema has been properly implemented. We don't recommend using null, cause we need some extra validating.